src/dataflow/src/render/upsert.rs

```diff
@@ -309,3 +320,125 @@ where
         (prepended, operator::empty(scope))
     }
 }
+
+pub fn no_arrangement_upsert<G, K, V>(
+    stream: &Stream<G, (SourceOutput<Vec<u8>, Vec<u8>>, Timestamp)>,
+    mut key_decoder_state: K,
+    mut value_decoder_state: V,
+) -> Stream<G, (Row, Timestamp, Diff)>
+where
+    G: Scope<Timestamp = Timestamp>,
+    K: DecoderState + 'static,
+    V: DecoderState + 'static,
+{
+    stream.unary_frontier(Pipeline, "LightUpsert", |_cap, _info| {
+        // to_send
+        // this is a map of (time) -> (capability, ((key) -> (value with max offset)))
+        let mut to_send = BTreeMap::<_, (_, HashMap<_, SourceData>)>::new();
+        let mut current_values = HashMap::new();
+        let mut vector = Vec::new();
+        let mut row_packer = repr::RowPacker::new();
+
+        move |input, output| {
+            // Digest each input, reduce by presented timestamp.
+            input.for_each(|cap, data| {
```

Comment:
i forget exactly when this applies but sometimes you just want to stash away inputs and then operate only on them to make sure the operator does an amount of work bounded by the inputs present when it was invoked. as it stands, this code doesn't do that but maybe it should? i honestly am not sure